cirq/ops/pauli_gates.py

```diff
@@ -119,6 +117,14 @@ def _from_json_dict_(cls, exponent, global_shift, **kwargs):
         assert exponent == 1
         return Pauli._XYZ[0]
 
+    @property
+    def basis(self):
```

Comment:
WDYT about moving this functionality into the `EigenGate` (where I think you may be able to benefit from the gate knowing the projectors onto their eigenspaces)?

Also, I imagine it may be useful sometimes to have a way to get *a* non-stabilizer state and something like `cirq.H.basis[+1]` is an obvious thing to try.

Reply:
You can get the projectors, yes but not the state vectors; and that's the important part. In particular, the state vector ndarrays are in the `_[_]EigenState` classes which are only implemented for `X`, `Y`, `Z`. Ie, we'd have to put another one in for `H`.

Reply:
This is such a good idea I already thought of it and the test is here! https://github.com/quantumlib/Cirq/pull/3171/files#diff-c5d792e951ba985282f75339939b6e14R60

(And indeed the projector is computed from the outer product of the state vector)

Reply:
Cool! Thanks! I expect over time we'll add more eigenbases and we'll want to test each for consistency with the projectors, so we should move this into `assert_eigengate_implements_consistent_protocols`.

Reply:
I mean I don't consider this a "protocol" at this point. After a quick look at the plumbing for `assert_eigengate_implements_consistent_protocols` it's not clear how to add this check other than shoe-horning it in for X, Y, Z gates. I find it much clearer to have this of the projector method alongside the projector method

Reply:
Fair enough. I think a protocol would be appropriate here (especially since most gates do not support `.basis`), but it's reasonable to say it's out-of-scope for this PR.

BTW: Should I have another look or do you plan some more changes here?

Reply:
Just added type annotations. Have a look!

Since most gates don't support basis doesn't that mean it's *not* appropriate for a protocol? I know there's no widely accepted definition for what makes a good Cirq protocol but usually it's "an abstract api with many fallback implementations" and I don't see that here.

The real sticking point is that you have to write down the state vector representations "by hand" somewhere

Reply:
Most gates don't provide their eigenbasis, but they do provide enough other information that it isn't hard to imagine various fallbacks that derive the eigenbasis *for certain types of gates*. Hence, a protocol. It might look something like this:

```
if the gate provides its eigenbasis explicitly then return that (e.g. X)
else if the gate is a tensor product of gates then return tensor product of their eigenbases (e.g. XX)
else if the gate is an EigenGate then use its projector (e.g. H)
else compute eigendecomposition or fail or whatever
```

Sure, we need to write down the eigenbasis somewhere for *some* gates, but not all. For other gates we can compute it from existing information (e.g. the projector for EigenGates) and we can compute it recursively (e.g. XX).

In any case, my point wasn't really about implementing a new protocol, but about testing consistency of eigenbases with projectors which you're doing in the test. That's good enough for now.